contracts/utils/Strings.sol

```diff
@@ -7,48 +7,101 @@ pragma solidity ^0.8.0;
  * @dev String operations.
  */
 library Strings {
-    bytes16 private constant _HEX_SYMBOLS = "0123456789abcdef";
+    bytes16 private constant _SYMBOLS = "0123456789abcdef";
     uint8 private constant _ADDRESS_LENGTH = 20;
 
     /**
      * @dev Converts a `uint256` to its ASCII `string` decimal representation.
      */
     function toString(uint256 value) internal pure returns (string memory) {
-        // Inspired by OraclizeAPI's implementation - MIT licence
-        // https://github.com/oraclize/ethereum-api/blob/b42146b063c7d6ee1358846c198246239e9360e8/oraclizeAPI_0.4.25.sol
+        unchecked {
+            uint256 length = 1;
 
-        if (value == 0) {
-            return "0";
-        }
-        uint256 temp = value;
-        uint256 digits;
-        while (temp != 0) {
-            digits++;
-            temp /= 10;
-        }
-        bytes memory buffer = new bytes(digits);
-        while (value != 0) {
-            digits -= 1;
-            buffer[digits] = bytes1(uint8(48 + uint256(value % 10)));
-            value /= 10;
+            // compute log10(value), and add it to length
+            uint256 valueCopy = value;
+            if (valueCopy >= 10**64) {
+                valueCopy /= 10**64;
+                length += 64;
+            }
+            if (valueCopy >= 10**32) {
+                valueCopy /= 10**32;
+                length += 32;
+            }
+            if (valueCopy >= 10**16) {
+                valueCopy /= 10**16;
+                length += 16;
+            }
+            if (valueCopy >= 10**8) {
+                valueCopy /= 10**8;
+                length += 8;
+            }
+            if (valueCopy >= 10**4) {
+                valueCopy /= 10**4;
+                length += 4;
+            }
+            if (valueCopy >= 10**2) {
+                valueCopy /= 10**2;
+                length += 2;
+            }
+            if (valueCopy >= 10**1) {
+                length += 1;
+            }
+            // now, length is log10(value) + 1
+
+            string memory buffer = new string(length);
+            /// @solidity memory-safe-assembly
+            assembly {
+                let ptr := add(buffer, add(32, length))
+                for {
+
+                } 1 {
+
+                } {
+                    ptr := sub(ptr, 1)
+                    mstore8(ptr, byte(mod(value, 10), _SYMBOLS))
+                    value := div(value, 10)
+                    if iszero(value) {
+                        break
+                    }
+                }
+            }
```

Comment:
If we declare `symbols` as a local variable before use, the compiler emits better code, loading the constant on the stack before the loop and then reusing that value, whereas using the state variable directly the constant is loaded on the stack on each iteration.

Using a while loop instead of an assembly for loop the gas costs are exactly the same and we get better readability. I didn't check the bytecode but I wouldn't be surprised if they're the same.

```suggestion
            uint ptr;
            assembly {
                ptr := add(buffer, add(32, length))
            }

            bytes16 symbols = _SYMBOLS;

            while (true) {
                ptr -= 1;
                assembly {
                    mstore8(ptr, byte(mod(value, 10), symbols))
                }
                value /= 10;
                if (value == 0) {
                    break;
                }
            }
```

Reply:
`ptr -= 1;` should be unchecked

Reply:
All of the function is in an `unchecked` block!

Reply:
Indeed the amount of assembly can be reduced with an exactly zero impact on gas. Good catch!

Copying of `_SYMBOLS` to a local variable increases the gas usage by over 17% for larger numbers. Why making a copy is supposed to reduce gas? Unlike managing yet another local variable on stack pushing a constant is a single opcode.

Reply:
I saw it reduce gas but didn't test large numbers. In retrospect I agree with the reasoning that managing a value on the stack will overall just add overhead.